qiskit/circuit/library/generalized_gates/permutation.py

```diff
@@ -77,17 +83,59 @@ def __init__(
             pattern = np.arange(num_qubits)
             rng.shuffle(pattern)
 
-        name = "permutation_" + np.array_str(pattern).replace(" ", ",")
+        # This is needed to support qasm()
+        self._qasm_name = "permutation__" + "_".join([str(n) for n in pattern]) + "_"
+        self._qasm_definition = None
 
-        circuit = QuantumCircuit(num_qubits, name=name)
+        super().__init__(name="permutation", num_qubits=num_qubits, params=[pattern])
 
-        super().__init__(num_qubits, name=name)
+    def __array__(self, dtype=None):
```

Comment:
Is there a test for this?

Reply:
Good point. I have added the missing tests for this and other new functionality in [e826085](https://github.com/Qiskit/qiskit-terra/pull/9157/commits/e8260856117a83db1221eda62e3c1a654450ac89).